collect/central_collector.go

```diff
@@ -478,12 +515,171 @@ func (c *CentralCollector) sendTraces(ctx context.Context) error {
+			traceIDs = append(traceIDs, traceID)
+			// if we reached the max count, we need to send
+			if len(traceIDs) >= maxCount {
+				send = true
+				maxTime = adjustMaxTime(maxTime, 0.8)
+			}
+			// we got one, so reset the ticker so it doesn't fire
+			ticker.Reset(maxTime)
+		case <-ticker.Chan():
+			// ticker fired, so send what we have
+			send = true
+			if len(traceIDs) < 2 {
+				if lessThanTwo {
+					maxTime = adjustMaxTime(maxTime, 1.2)
+				}
+				lessThanTwo = true
+			} else {
+				lessThanTwo = false
+			}
+		}
+		// if we need to send, do so
+		if send && len(traceIDs) > 0 {
+			// we get back the ones that didn't process immediately
+			notready := f(traceIDs)
+			if len(notready) > maxCount/2 {
+				// if we have more than half the batch returned, we may be having trouble keeping up
+				// so we should back off a bit
+				maxTime = adjustMaxTime(maxTime, 1.2)
+			}
+			traceIDs = append(traceIDs[:0], notready...)
+			send = false
+		}
+	}
+}
+
+// keepTraces needs to retrieve the status of the traces from the central store
+// so that it can attach the metadata as needed
+func (c *CentralCollector) keepTraces(ids []string) []string {
+	ctx, span := otelutil.StartSpanWith(context.Background(), c.Tracer, "CentralCollector.keepTraces", "num_ids", len(ids))
+	defer span.End()
+	statuses, err := c.Store.GetStatusForTraces(ctx, ids, centralstore.DecisionKeep)
+	if err != nil {
+		c.Logger.Error().Logf("error getting status for traces: %s", err)
+	}
+
+	idset := generics.NewSet(ids...)
+	for _, status := range statuses {
+		c.sendSpans(status)
+		c.SpanCache.Remove(status.TraceID)
+		idset.Remove(status.TraceID)
+		c.Metrics.Increment("collector_keep_trace")
+	}
+	return idset.Members()
+}
+
+// dropTraces doesn't need to retrieve the status of the traces from the central store
+// because it's only removing the traces from the cache
+func (c *CentralCollector) dropTraces(ids []string) []string {
+	_, span := otelutil.StartSpanWith(context.Background(), c.Tracer, "CentralCollector.dropTraces", "num_ids", len(ids))
+	defer span.End()
+	for _, traceID := range ids {
+		c.SpanCache.Remove(traceID)
+		c.Metrics.Increment("collector_drop_trace")
+	}
+	return nil
+}
+
+// The cleanup task is responsible for removing traces from the cache that have
+// been around for too long. This is a hedge against process restarts where the
+// gossiped message was lost or happened before we existed.
+func (c *CentralCollector) cleanup() error {
+	return c.cleanupCycle.Run(context.Background(), func(ctx context.Context) error {
+		c.cleanupTraces(ctx)
+		c.Metrics.Increment("collector_cleanup_runs")
+		return nil
+	})
+}
+
+// Cleanup traces asks for old (expired) trace IDs from the cache and then gets the status
+// (either keep or drop) from the central store, and then dispatches them appropriately.
+// In a stable refinery cluster, this should be a no-op.
+func (c *CentralCollector) cleanupTraces(ctx context.Context) {
```

Comment:
Does this cleanup process have an effect on the spans processed metrics? Should we be incrementing that metric since we are "reprocessing" the span?